File: Go/Rofi-hkey/json_editor.go
package main

import (
    "rofi-hkey/rofi"
    "os"
    "fmt"
    "strings"
    "io/ioutil"
    "encoding/json"
)

func addHkey() {
    newHkey := nameHkey("")
    if newHkey == "q" {
        return
    }

    newCommand := getCommand("")
    if newCommand == "q" {
        return
    }

    newDescription := getDescription("")
    if newDescription == "q" {
        return
    }

    if !confirmation("Add", newHkey, newCommand, newDescription) {
        return
    }

    hkeys[newHkey] = []string{newCommand, newDescription}
    go updateHkeysJson()
    rofi.MenuMessage = " -mesg \"'" + newHkey + "' Added!\""
}

func removeHkey() {
    hkey := hkeySearch("Search Hkey to Remove")
    if hkey == "q" {
        return
    }
    command := hkeys[hkey][0]
    description := hkeys[hkey][1]

    if !confirmation("Remove", hkey, command, description) {
        return
    }

    delete(hkeys, hkey)
    go updateHkeysJson()
    rofi.MenuMessage = " -mesg \"'" + hkey + "' Removed!\""
}

func editHkey() {
    hkey := hkeySearch("Search Hkey to Edit")
    if hkey == "q" {
        return
    }
    command := hkeys[hkey][0]
    description := hkeys[hkey][1]
    rofi.MenuMessage = ""

    var section string
    sectionEdit := func(sectionOld *string, sectionFunc func(string) string) {
        menuMessagePrefix := "Current " + section + ": " + *sectionOld + "\n"
        sectionNew := sectionFunc(menuMessagePrefix)
        if sectionNew == "q" {
            return
        } else if sectionNew == *sectionOld {
            rofi.MenuMessage = " -mesg \"Nothing has been changed...\""
            return
        }

        if section == "Hkey" {
            delete(hkeys, *sectionOld)
        }

        rofi.MenuMessage = " -mesg \"" + section + " changed from '" + *sectionOld + "' to '" + sectionNew + "'\""
        *sectionOld = sectionNew
        hkeys[hkey] = []string{command, description}
        go updateHkeysJson()
    }

    for {
        prompt := "Edit what?"
        dmenu := []string {"Hkey: " + hkey, "Command: " + command, "Description: " + description, "Quit"}
        section = rofi.CustomDmenu(prompt, dmenu, false)
        if strings.Contains(section, ":") {
            section = strings.Split(section, ":")[0]
        }

        switch(section) {
        case "Hkey":
            sectionEdit(&hkey, nameHkey)
        case "Command":
            sectionEdit(&command, getCommand)
        case "Description":
            sectionEdit(&description, getDescription)
        case "Quit":
            rofi.MenuMessage = " -mesg \"Exited edition mode\""
            return
        }
    }
}

func hkeySearch(prompt string) string {
    rofi.MenuMessage = ""
    hkeySearchLoop:for {
        dmenu := createHkeysArray()
        userInput := rofi.CustomDmenu(prompt, dmenu, true)

        if userInput == "q" {
            return "q"
        }

        for _, rkeyEntry := range rkeys {
            rkey := rkeyEntry[0]
            if userInput == rkey {
                description := rkeyEntry[1]
                rofi.MenuMessage = " -mesg \"'" + rkey + "' is a reserved key used to: " + description + "\""
                continue hkeySearchLoop
            }
        }

        return userInput
    }
}

func nameHkey(menuMessagePrefix string) string {
    rofi.MenuMessage = " -mesg \"" + menuMessagePrefix + "[If you'll run the command with an input (e.g.: google 'input') make sure the hkey has a space at the end, otherwise don't leave a space]\""

    nameHkeyLoop:for {
        newHkey := rofi.SimplePrompt("Enter the New Hkey")

        if newHkey == "q" || newHkey == "" {
            rofi.MenuMessage = " -mesg \"Aborted...\""
            return "q"
        }

        for hkey := range hkeys {
            if newHkey == hkey {
                description := hkeys[newHkey][1]
                rofi.MenuMessage = " -mesg \"'" + newHkey + "' already exists to: '" + description + "'\""
                continue nameHkeyLoop
            }
        }

        for _, rkeyEntry := range rkeys {
            rkey := rkeyEntry[0]
            if newHkey == rkey {
                description := rkeyEntry[1]
                rofi.MenuMessage = " -mesg \"'" + rkey + "' is a reserved key used to: " + description + "\""
                continue nameHkeyLoop
            }
        }

        return newHkey
    }
}

func getCommand(menuMessagePrefix string) string {
    if menuMessagePrefix != "" {
        rofi.MenuMessage = " -mesg \"" + menuMessagePrefix + "\""
    } else {
        rofi.MenuMessage = ""
    }
    newCommand := rofi.SimplePrompt("Enter the full command to link to the Hkey")

    if newCommand == "q" || newCommand == "" {
        rofi.MenuMessage = " -mesg \"Aborted...\""
        return "q"
    }

    return newCommand
}

func getDescription(menuMessagePrefix string) string {
    if menuMessagePrefix != "" {
        rofi.MenuMessage = " -mesg \"" + menuMessagePrefix + "\""
    } else {
        rofi.MenuMessage = ""
    }
    for {
        newDescription := rofi.SimplePrompt("Enter a description of the command")

        if newDescription == "q" || newDescription == "" {
            rofi.MenuMessage = " -mesg \"Aborted...\""
            return "q"
        } else if len(newDescription) > 60 {
            rofi.MenuMessage = " -mesg \"Description is too long...\""
            continue
        }

        newDescriptionArray := strings.Split(strings.TrimSpace(newDescription), " ")
        newDescription = ""
        // Faço desta forma pq podem haver siglas (com todas as letras maiúsculas)
        // Melhorar isto
        for i, word := range newDescriptionArray {
            word = strings.ToUpper(word[:1]) + word[1:]
            newDescription += word
            if i < len(newDescriptionArray) - 1 {
                newDescription += " "
            }
        }

        return strings.TrimSuffix(newDescription, " ")
    }
}

func confirmation(mode string, hkey string, command string, description string) bool{
    prompt := mode + " this entry?"
    rofi.MenuMessage = " -mesg \"Hkey: " + hkey + "\nCommand: " + command + "\nDescription: " + description + "\""
    dmenu := []string {"Yes", "No"}

    confirmation := rofi.CustomDmenu(prompt, dmenu, false)
    if confirmation == "Yes" {
        return true
    } else {
        rofi.MenuMessage = " -mesg \"Aborted...\""
        return false
    }
}

func updateHkeysJson() {
    file, err := json.MarshalIndent(hkeys, "", "    ")

	err = ioutil.WriteFile(hkeysPath, file, 0644)
    if err != nil {
        fmt.Println(err)
        os.Exit(1)
    }
}

File: Python/Hkey/json_editor.py
#!/usr/bin/python3

import json


class JsonEditor:
    def __init__(
        self, hkeys: dict[str, tuple], rkeys: dict[str, tuple]
    ) -> None:
        self.hkeys = hkeys
        self.rkeys = rkeys

    def add_hkey(self) -> None:
        print("\nAdding mode")
        self.hkey = self.name_hkey()
        if self.hkey in ("", "q"):
            return

        self.cmd = self.get_command()
        if self.cmd in ("", "q"):
            return

        self.desc = self.get_description()
        if self.desc in ("", "q"):
            return

        self.new_sh = self.get_shell_session()
        if self.new_sh in ("", "q"):
            return

        if not self.confirmation("Add"):
            return

        self.hkeys[self.hkey] = self.cmd, self.desc, self.new_sh
        self.update_hkeys()
        print("Added!\n")

    def remove_hkey(self) -> None:
        print("\nRemoving mode")
        self.hkey = self.hkey_search()
        if self.hkey in ("", "q"):
            return

        self.cmd, self.desc, self.new_sh = self.hkeys[self.hkey]

        if not self.confirmation("Remove"):
            return

        self.hkeys.pop(self.hkey)
        self.update_hkeys()
        print("Removed!\n")

    def edit_hkey(self) -> None:
        print("\nEditing mode")
        self.hkey = self.hkey_search()
        if self.hkey in ("", "q"):
            return
        self.cmd, self.desc, self.new_sh = self.hkeys[self.hkey]
        section_opts = {
            "0": ["Hkey", self.hkey, self.name_hkey],
            "1": ["Command", self.cmd, self.get_command],
            "2": ["Description", self.desc, self.get_description],
            "3": ["Shell", self.new_sh, self.get_shell_session],
        }

        while True:
            [
                print(
                    f"[{opt}] {section_opts[opt][0]}: '{section_opts[opt][1]}'"
                )
                for opt in section_opts.keys()
            ]

            section = input("Edit what? (Enter 'q' to exit edition mode): ")
            if section == "q":
                print("Exited edition mode")
                return
            elif section not in section_opts:
                continue

            print(
                f"Current {section_opts[section][0]}: "
                f"'{section_opts[section][1]}'"
            )
            edition = section_opts[section][2]()

            if edition in ("", "q"):
                return
            elif edition == section_opts[section][1]:
                print("Nothing has been changed...")
                continue

            print(
                f"{section_opts[section][0]} changed from "
                f"'{section_opts[section][1]}' to '{edition}'"
            )
            section_opts[section][1] = edition
            self.hkeys.pop(self.hkey)
            self.hkeys[section_opts["0"][1]] = (
                section_opts["1"][1],
                section_opts["2"][1],
                section_opts["3"][1],
            )

            self.update_hkeys()
            print("Edition Saved!")

    def hkey_search(self) -> str:
        while True:
            search_hkey = input("Enter hkey to search: ")
            if search_hkey in ("", "q"):
                print("Aborting...")
                return search_hkey
            elif search_hkey in self.hkeys:
                return search_hkey
            elif search_hkey in self.rkeys:
                print(f"'{search_hkey}' is a reserved key")
            else:
                print(f"'{search_hkey}' not found")

    def name_hkey(self) -> str:
        print(
            "\n[If you'll run the command with an input (e.g.: google "
            "<input>) make sure the hkey has a space at the end, "
            "otherwise don't leave a space]"
        )

        while True:
            new_hkey = input("\nEnter the new hkey: ")
            if new_hkey in ("", "q"):
                print("Aborting...")
            elif new_hkey in self.hkeys:
                print(
                    f"'{new_hkey}' already exists for the following "
                    f"command: '{self.hkeys[new_hkey][0]}'"
                )
                continue
            elif new_hkey in self.rkeys:
                print(f"'{new_hkey}' is a reserved key...")
                continue
            return new_hkey

    def get_command(self) -> str:
        cmd = input("Enter the full command to link to the hkey: ")
        if cmd in ("", "q"):
            print("Aborting...")
        return cmd

    def get_description(self) -> str:
        while True:
            new_desc = input("Enter a description of the command: ")
            if new_desc in ("", "q"):
                print("Aborting...")
            elif len(new_desc) > 80:
                print("Description is too long...")
                continue
            return new_desc

    def get_shell_session(self) -> str:
        new_sh = input(
            ":: When launched, should the command be "
            "started on a new shell session? [y/N] "
        )
        if new_sh.lower() in ("", "n"):
            return "Same Session"
        elif new_sh.lower() == "y":
            return "New Session"
        else:
            print("Aborting...")
            return "q"

    def confirmation(self, mode) -> bool:
        ans = input(
            f"\nHkey: {self.hkey}\nCommand: {self.cmd}\n"
            f"Description: {self.desc}\nShell: "
            f"{self.new_sh}\n:: {mode} this entry? [Y/n] "
        )

        if ans.lower() in ("", "y"):
            return True
        else:
            print("Aborting...")
            return False

    def update_hkeys(self) -> None:
        hkeys_path = "hkeys_path"
        self.hkeys = dict(sorted(self.hkeys.items()))
        with open(hkeys_path, "w") as hk:
            json.dump(self.hkeys, hk, indent=4)
